### PR TITLE
feat: add health check for minimal memory / disk size

### LINKED ISSUE #6467: Report when a control plane node does not meet requirements
## Feature Request
Have talosctl health report if minimum requirements are not met.
### Description
Currently, if you fire up a node with 1GB of RAM and try to make it a control plane, the node just gets stuck.
Talosctl reports:
```
waiting for all nodes to finish boot sequence: ...
waiting for all nodes to finish boot sequence: nodes ["45.32.84.84"] are still in boot sequence
```
and just sits there. Logs show
```
 "k8s.KubeletStaticPodController", "error": "error refreshing pod status: error fetching pod status: an error on the server (\"Authorization error (user=apiserver-kubelet-client, verb=get, resource=nodes, subresource=proxy)\") has prevented the request from succeeding"}
```

It would be helpful if talosctl health reported "Control plane nodes require at least 2GB of memory", if they do not have that.
Also would be good if `talosctl apply-config --file controlplane.yaml` did the same thing...

### PR DESCRIPTION
**Pull request by DmitriyMV**
This PR adds two additional checks which are performed during boot sequence and in `talosctl health`. They ensure that nodes have enough memory and disk.

- Boot check will print a warning if memory / disk size is not sufficient.
- Health check will fail if memory / disk size is not sufficient.

Closes #6467

Signed-off-by: Dmitriy Matrenichev <dmitry.matrenichev@siderolabs.com>

### CODE REVIEW

File: pkg/cluster/check/nodes.go [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+		}
+
+		hostname := msg.Metadata.Hostname
+
+		typ, ok := nodes[hostname]
+		if !ok {
+			return fmt.Errorf("unexpected node %q in response", hostname)
+		}
+
+		if msg.Metadata.Error != "" {
```
Comment (smira) [inline pkg/cluster/check/nodes.go:57]: this is not needed, `Memory` function of the client does error unwrapping removing any `Metadata.Error` responses and pushing them to `err error` response

File: pkg/cluster/check/nodes.go [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+			resultErr = multierror.Append(resultErr, err)
+
+			continue
+		}
+
+		if totalMemory := msg.Meminfo.Memtotal * humanize.KiByte; totalMemory < minimum {
+			resultErr = multierror.Append(
+				resultErr,
+				fmt.Errorf(
+					"node %q has too small memory: %d MiB, minimum is %d MiB",
```
Comment (smira) [inline pkg/cluster/check/nodes.go:77]: @rsmitty I think this message should be rephrased, `too little memory` is more correct, but sounds weird?

File: pkg/cluster/check/nodes.go [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+		return nil, fmt.Errorf("no nodes with mounts found")
+	}
+
+	nodesMnts := map[string][]mntData{}
+
+	for _, msg := range diskResp.Messages {
+		switch {
+		case msg.Metadata == nil:
+			return nil, fmt.Errorf("no metadata in response")
+		case msg.Metadata.Error != "":
```
Comment (smira) [inline pkg/cluster/check/nodes.go:219]: same thing with error here

File: pkg/cluster/check/nodes.go (hunk 2 above)
Comment (rsmitty) [reply pkg/cluster/check/nodes.go:77]: Maybe "node %q does not meet memory requirements..."?

File: pkg/minimal/limits.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+// Memory returns the minimal/recommended amount of memory required to run the node.
+func Memory(typ machine.Type) (minimum, recommended uint64, err error) {
+	// We remove 100 MiB from the recommended memory to account for the kernel
+	switch typ { //nolint:exhaustive
+	case machine.TypeControlPlane, machine.TypeInit:
+		minimum = 2*humanize.GiByte - 100*humanize.MiByte
+		recommended = 4*humanize.GiByte - 100*humanize.MiByte
+
+	case machine.TypeWorker:
+		minimum = 2*humanize.GiByte - 100*humanize.MiByte
```
Comment (smira) [inline pkg/minimal/limits.go:25]: same settings?

Comment (DmitriyMV) [reply pkg/minimal/limits.go:25]: Copy paste error.

File: pkg/cluster/check/nodes.go (hunk 3 above)
Comment (DmitriyMV) [reply pkg/cluster/check/nodes.go]: Fixed!

File: pkg/cluster/check/nodes.go (hunk 2 above)
Comment (DmitriyMV) [reply pkg/cluster/check/nodes.go]: Took @rsmitty suggestion :)

File: pkg/cluster/check/nodes.go (hunk 1 above)
Comment (DmitriyMV) [reply pkg/cluster/check/nodes.go]: Fixed!

File: internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_sequencer_tasks.go [hunk 5]
```diff
@@ -26,8 +26,10 @@ import (
 	cgroupsv2 "github.com/containerd/cgroups/v2"
 	"github.com/cosi-project/runtime/pkg/resource"
 	"github.com/cosi-project/runtime/pkg/state"
-	multierror "github.com/hashicorp/go-multierror"
+	"github.com/dustin/go-humanize"
+	"github.com/hashicorp/go-multierror"
 	"github.com/opencontainers/runtime-spec/specs-go"
+	procfs2 "github.com/prometheus/procfs"
```
Comment (utkuozdemir) [inline internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_sequencer_tasks.go:32]: maybe use a better name like `promprocfs` or something?

Comment (DmitriyMV) [reply internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_sequencer_tasks.go]: Now its caled pprocfs 😄

File: pkg/cluster/check/nodes.go [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+}
+
+type mountData struct {
+	Source string
+}
+
+// ErrOldTalosVersion is returned when the node is running an old version of Talos.
+var ErrOldTalosVersion = fmt.Errorf("old Talos version")
+
+func getEphermalPartitionData(ctx context.Context, state state.State, nodeIP string) (mountData, error) {
```
Comment (utkuozdemir) [inline pkg/cluster/check/nodes.go:180]:
```suggestion
func getEphemeralPartitionData(ctx context.Context, state state.State, nodeIP string) (mountData, error) {
```

typo

File: internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_sequencer_tasks.go [hunk 7]
```diff
@@ -681,6 +684,82 @@ func ValidateConfig(seq runtime.Sequence, data interface{}) (runtime.TaskExecuti
...
+		diskSize, err := disk.Size()
+		if err != nil {
+			return fmt.Errorf("failed to get ephemeral disk size: %w", err)
+		}
+
+		if minimum := minimal.DiskSize(); diskSize < minimum {
+			log.Println("WARNING: disk size is less than recommended")
+			log.Println("WARNING: Talos may not work properly")
+			log.Println("WARNING: minimum recommended disk size is", minimum/humanize.MiByte, "MiB")
+			log.Println("WARNING: current total disk size is", diskSize/humanize.MiByte, "MiB")
```
Comment (utkuozdemir) [inline internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_sequencer_tasks.go:753]: I am overall confused with what we want to do if the checks a. fail b. conditions not met.

Here we just print warnings, but if we fail to read the disk size for example, then we return an error. Shouldn't we be consistent here and either warn or fail on all cases?

Question also applies to the health checks - there we fail explicitly, but here we only warn. What's the expected behavior here?

**Review by smira — APPROVED** [review]

**Pull request merged**